Refactor to conform with ES-Shim API Contract

JakeChampion opened the pull request:
With the two together, it is impossible to test whether the module
works in engines which partially support Array.prototype.find.

With the two separated, we can test in all engines and we can more
easily use this polyfill inside of the [Financial Times's polyfill-service](https://github.com/Financial-Times/polyfill-service).

JakeChampion commented:
CI doesn't seem to be set-up for this repository anymore, I have ran the tests in multiple versions of NodeJS myself (Versions 0.10/0.11/0.12/4/5). [The results are in a gist](https://gist.github.com/JakeChampion/e330205c7bb968a6cf5f). All the tests are passing.

ljharb commented on array-prototype-find.js:
```diff
@@ -1,9 +1,7 @@
-  if (Array.prototype.find) return;
-
-  var find = function(predicate) {
+  Array.prototype.find = function find (predicate) {
```
it'd be nice to change this file so the implementation could be used without the shimming behavior - ie, move the shim assignment into `detect.js`?

JakeChampion replied:
Hmm, I'm wanting to get the shimming behaviour without the detect. What do you think about having three files? `detect.js`, `polyfill.js`, `shim.js`?

ljharb replied:
yes, let's please do that. see also: https://github.com/es-shims/es-shim-api

JakeChampion replied:
That looks like a good specification to follow. Thanks for the link.

JakeChampion commented:
@ljharb I've modified the project structure to now follow the ES-Shim API Contract. Hopefully I've followed it correctly.

ljharb commented on index.js:
```diff
@@ -1,32 +1,22 @@
+	shim: shim
+});
+
+module.exports = boundFindShim;
```
trailing newline (you may want to configure your editor :-))

JakeChampion replied:
I don't have trailing new lines on by default however, my editor will follow a project's style guide via [editorconfig](http://editorconfig.org/) and [eslint](http://eslint.org/). I can submit a PR with the styles you are wanting me to follow if you want?

ljharb replied:
sure, an editorconfig wouldn't hurt

ljharb commented on polyfill.js:
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+module.exports = function getPolyfill() {
+	return Array.prototype.find || require('./implementation');
```
There is a bug in early implementations of `Array#find` that this should be detecting - https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L1432-L1440

JakeChampion replied:
Good catch!

ljharb commented on shim.js:
```diff
@@ -0,0 +1,12 @@
+module.exports = function shimArrayPrototypeFind() {
+	var polyfill = getPolyfill();
+	if (Array.prototype.find !== polyfill) {
+		define(Array.prototype, { find: polyfill });
```
`define` won't override existing properties unless there's a corresponding "should override" function - ie, here you want `define(Array.prototype, { find: polyfill }, { find: function () { return true; })`. Alternatively you could move the "if" condition into the predicate function.

JakeChampion replied:
Thanks for spotting this :-)

ljharb commented:
This looks great!

@paulmillr will have to enable travis-ci, however, since non-organization-owned repos can't enable admin  access for others.

ljharb commented:
It would also be ideal to add `@es-shims/api` as a dev dependency, and use it to run basic conformance tests - like https://github.com/es-shims/String.prototype.trim/blob/master/package.json#L20 for example.

JakeChampion commented:
@ljharb Comments have been addressed :-)

ljharb commented on implementation.js:
```diff
@@ -0,0 +1,19 @@
+'use strict';
+var ES = require('es-abstract/es6');
+
+module.exports = function find (predicate) {
```
please remove the space between the function's name and the argument list

ljharb commented:
Previously, this project used 2-space indentation - most of your added files use tabs. I prefer tabs, so either one is fine, but we should make sure the whole project is consistent one way or another.

JakeChampion commented:
@ljharb Migrated to tabs, added the change to `.editorconfig` for JS and JSON files. I've also removed the space between the function name and argument list. May I ask if you are following a certain community style guide? I could include it in a subsequent pull request if you think that would be helpful for contributors.

ljharb commented on implementation.js:
```diff
@@ -0,0 +1,19 @@
+var ES = require('es-abstract/es6');
+
+module.exports = function find(predicate) {
+  var list = ES.ToObject(this);
```
this file still uses spaces

ljharb commented on tests/test.js:
```diff
@@ -81,11 +80,32 @@ var runTests = function () {
+describe('shim', function() {
+	find.shim();
+	var implementation = Array.prototype.find;
+	console.log(implementation);
```
remove this?

ljharb commented on tests/test.js:
```diff
@@ -81,11 +80,32 @@ var runTests = function () {
+    Object.prototype[1] = 42;
+    runTests(implementation);
+    delete Object.prototype[1];
+  });
```
still some spaces here too

ljharb commented:
@JakeChampion somewhere between https://github.com/ljharb/eslint-config and https://github.com/airbnb/javascript - but "no space after the function name" and "always add a trailing newline" is pretty universal afaict.

JakeChampion commented:
@ljharb Updated

ljharb commented:
This LGTM - I'm going to give it a few days before merging though.

JakeChampion commented:
All good with me, there's no rush with this work :-)

I'm happy to squash the commits if you would like a cleaner history

ljharb commented:
Absolutely that would be appreciated - not down to 1, of course, but combining all the conceptually related ones so there's a smaller number of conceptually atomic commits would be great :-)

JakeChampion commented:
I've squashed conceptually related commits together. If you think more can be squashed together I will happily do them also.

paulmillr commented:
Thanks Jake.
